htpy/__init__.py

```diff
@@ -130,11 +135,26 @@ class ContextProvider(t.Generic[T]):
     value: T
     node: Node
 
+    @deprecated(
+        "iterating over a context provider is deprecated and will be removed in a future release. "
+        "Please use the context_provider.iter_chunks() method instead."
+    )  # pyright: ignore [reportUntypedFunctionDecorator]
     def __iter__(self) -> Iterator[str]:
-        return iter_node(self)
+        return self.iter_chunks()
+
+    def __str__(self) -> _Markup:
+        return _as_markup(self)
 
-    def __str__(self) -> str:
-        return render_node(self)
+    __html__ = __str__
+
+    def iter_chunks(self) -> Iterator[str]:
+        return _iter_chunks_node(self, {})
```

Comment:
I guess we could skip a level in the call stack here and go directly for:

```suggestion
        return self._iter_chunks_context({})
```

Same for `ContextConsumer`, `BaseElement` and `Fragment`.

Reply:
very true, changed now!

we could also change `def iter_chunks(self) -> Iterator[str]:` to something like `def iter_chunks(self, context: dict[Context[t.Any], t.Any]=None) -> Iterator[str]:` and get rid of _iter_chunks_context altogether. then implementing Renderable would just need to implement public methods.

to make anything useful a custom implementation would still need to call into _iter_chunks_node (which we could rename to `iter_chunks_node` or something at that point. 

but maybe it is nicer to keep it as _iter_chunks_context for now and then see what other uses that comes up

Reply:
I went ahead and changed `iter_chunks()` to accept the context dict and got rid of `._iter_chunks_context()`. So now it is completely possible to implement anything outside of htpy by implementing `Renderable`. `_iter_chunks_node` is still private but easy to access via a fragment if one needs that from a custom Renderable.